source/common/http/http2/codec_impl.cc

```diff
@@ -537,7 +537,16 @@ void ConnectionImpl::ServerStreamImpl::resetStream(StreamResetReason reason) {
     buffer_memory_account_->clearDownstream();
   }
 
-  StreamImpl::resetStream(reason);
+  if (buffer_memory_account_ && buffer_memory_account_->sawEnvoyStreamComplete()) {
```

Comment:
So the underlying problem here, is that the filter manager believes the stream to be complete, and the codec does not?  I think we don't want the buffer accounting to be the entity handling this disconnect because there could be other entities with this lifetime confusion.

either entities which can down-call a reset should be explicitly made aware of a stream being complete, or the filter manager should handle reset being called after it thinks that the stream is complete.  Make sense?
cc @snowp for second opinion

Reply:
The Envoy level stream object (`ConnectionManagerImpl::ActiveStream` is complete) and so the filter manager also thinks the stream is complete, but the codec doesn't as you said as it has yet to call `onStreamClose`. 

AFAIK, the only entities that'll try to reset the stream at this point is either the `reset_high_memory_stream` action or the `OnPendingFlushTimer` though in the latter case the codec will know the stream has ended.

Reply:
there's also cases in envoy mobile where we've had a reset-after-fin (Snow has reviewed) so again I'd rather make this a bit more resistant